internal/controllers/hostedclusters/hostedcluster_controller.go

```diff
@@ -0,0 +1,108 @@
+package hostedclusters
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	corev1alpha1 "package-operator.run/apis/core/v1alpha1"
+)
+
+type HostedClusterController struct {
+	client           client.Client
+	log              logr.Logger
+	scheme           *runtime.Scheme
+	newHostedCluster hostedClusterFactory
+	image            string
+}
+
+func NewHostedClusterController(
+	c client.Client, log logr.Logger, scheme *runtime.Scheme, image string,
+) *HostedClusterController {
+	controller := &HostedClusterController{
+		client:           c,
+		log:              log,
+		scheme:           scheme,
+		newHostedCluster: newHostedCluster,
+		image:            image,
+	}
+	return controller
+}
+
+func (c *HostedClusterController) Reconcile(
+	ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := c.log.WithValues("HostedCluster", req.String())
+	defer log.Info("reconciled")
+	ctx = logr.NewContext(ctx, log)
+	hostedCluster := c.newHostedCluster()
+	if err := c.client.Get(ctx, req.NamespacedName, hostedCluster.ClientObject()); err != nil {
+		// Ignore not found errors on delete
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	desiredPackage := c.desiredPackage(hostedCluster)
+	// Can't use controllerutil.SetControllerReference because the HostedClusterType isn't in Scheme,
+	setControllerReference(hostedCluster.ClientObject(), desiredPackage)
```

Comment:
`HostedCluster` does not need to be part of the scheme, controllerutil.SetControllerReference uses apituil.GVKForObject, which is interrogating the Unstructured object itself for Kind and GroupVersion :)
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.13.1/pkg/client/apiutil/apimachinery.go#L76

Reply:
Ohhh awesome! Thanks! :)